File: Code/Python/Python_practice/binary_search.py

def sorted_array(array, key, start = 0, end=10):

    # base case of recursion
    if(len(array) == 0):
        return "not found"

    # finding the midpoint of the array
    mid = (start + end) // 2

    # if element found returning the place of the 
    if(array[mid] == key):
        return f"found at {mid}"
    
    # checking if the key is less then the element in the mid
    if(array[mid] > key):

        # if the key is less than the element in the mid
        # search for the left part of the array
        return sorted_array(array, key, start=0, end=mid-1)
    
    # checking if the key is greater than the element in the mid
    else:

        # if the key is greater than the element in the mid
        # search for the right part of the array
        return sorted_array(array, key, start=mid+1, end=len(array)-1)
        
array = [1,2,3,4,5,6,7,8,9]
print(sorted_array(array, 9))

for x in range(10):
    print(sorted_array(array, x))

File: Data_structures/Search/binary_search.c
#include <stdio.h>
#include <conio.h>


// 1,2,3,4,5

// algorithm on binary search

// 1) search for the number 4

// 2) pickup the center element of the array as key

// 3) if the center part of the array is less than 4
// then don't look into left part of array anymore

// 4) if the center part of the array is greater than 4
// then don't look into right part of the array anymore

// 5) continue this process till the element is found

int array[5] = {1,2,3,4,5};

int binary_search(int *array, int key, int front, int last)
{
    int mid;

    while(!(front > last))
    {
        // finding the key element which is at the center of the array
        mid = (front + last)/2;

        // checking if element is found
        if(array[mid] == key)
        {
            return mid;
        }

        if(array[mid] < key)
        {
            front = mid + 1;
        }
        else
        {
            last = mid - 1;
        }
    }

}

void main()
{

    int i;
    
    // asking user element to be searched

    printf("%d",binary_search(array, key, 0, 5-1));
    
    getch();

}
